### PR TITLE
delegate events to matching child elements if selector is present

### PR DESCRIPTION
**Pull request by jameslaneconkling**

### CODE REVIEW

**Comment by jameslaneconkling** [comment]
From #4 :
> This looks great! Couple things.
>
> Send this as a pull, much easier to collaborate / annotate lines that way.
>
> listener should be called with the view's this context instead of node. I know we're balancing D3 conventions and Backbone conventions but let's try to fit with most assumption made by Views.
>
> We need some way to set the delgateTarget/bubblingTarget of the Event to tell the listener what element it was originally delegated to (though see akre54/Backbone.NativeView#16 for some issues around naming. Any thoughts?)
> 
> Make sure to set d3.event (and cache the old one). Many of D3's methods rely on it being set.
> 
> Do we need namespace now? Mostly that's just a convention in Backbone to make undelegateEvents work easier. If we're binding our own events it shouldn't be necessary.
> 
> el.matches is pretty easy to pull out for maximum browser coverage
> 
> Otherwise this looks great. Let's add some tests and get this merged in.

**Comment by jameslaneconkling** [comment]
re: point 3: spent some time looking into this but didn't come up with anything yet.  Will have to do a bit more research on what's needed here.

Couldn't get the tests to pass yet.  Will take another look later today.  Post any feedback, if you have any.

File: backbone.d3view.js [hunk 1]
```diff
@@ -75,31 +75,54 @@
...
-      // The `event` object is stored in `d3.event` but Backbone expects it as
-      // the first argument to the listener.
-      el.on(eventName + namespace, function() {
-        var args = slice.call(arguments);
-        args.unshift(d3.event);
-        listener.apply(this, args);
-      });
+      handlers.push({selector: selector, listener: listener});
+
+      var ElementProto = (typeof Element !== 'undefined' && Element.prototype) || {};
```
Comment (akre54) [inline backbone.d3view.js]: pull this into the outer closure, along with the rest of the helpers. no need to recreate this on each call.

File: backbone.d3view.js [hunk 2]
```diff
@@ -75,31 +75,54 @@
...
+          return;
+        }
+        while(node && node !== el){
+          if(matchesSelector.call(node, selector)){
+            listener.apply(this, [d3.event, node.__data__, idx++]);
+          }
+          node = node.parentNode;
+        }
+        d3.event = o;
+      }.bind(this));
```
Comment (akre54) [inline backbone.d3view.js]: No need for `bind`, put a `var view = this` above the `addEventListener`. Much faster and more legible.

File: backbone.d3view.js [hunk 3]
```diff
@@ -130,7 +153,7 @@
 
       for (var eventName in map) {
         _.each(map[eventName], function(handler) {
-          removeEvent(this.d3el, eventName, handler.selector, handler.namespace);
+          removeEvent(this.d3el, eventName, handler.selector);
```
Comment (akre54) [inline backbone.d3view.js]: `removeEvent` should simply be a call to `removeEventListener`, no? We're not using D3's event management at all...

**Comment by akre54** [comment]
Pulled in manually to master, which also removes the Underscore dependency. It could use some tests for delegation behavior (to truly test that we've fixed it), but otherwise awesome work. Thanks for your help.

**Comment by jameslaneconkling** [comment]
@akre54 thanks for pulling this in.  Really interested to see where this could go--their data-centric philosophy make BB and D3 a natural pair.

Will keep my eyes on this.  Might make a PR to expand the documentation (if you'd be interested) once I get a little deeper into [the project I'm working on](https://github.com/jameslaneconkling/happyClassroom).

**Comment by akre54** [comment]
Oh cool, looks like a great tool. Keep me posted how it goes.

**Pull request merged**